crates/refunder/src/submitter.rs

```diff
@@ -0,0 +1,225 @@
+                    // send.
+                    tracing::debug!(?err, "transaction already mined");
+                } else if TX_ALREADY_KNOWN.iter().any(|msg| err.contains(msg)) {
+                    // This case means that the node is already aware of the tx
+                    // This can only happen after restarts, as usually we would always increase
+                    // the gas price compared to previous tx.
+                    // In this situation, we restart the submission loop with double the
+                    // previous gas. Due to the nature of EIP1559 this will not increase the
+                    // cost of the tx
+                    // todo: Finding last tx and its gas price would be nicer
+                    self.gas_price_of_last_submission = Some(max_fee_per_gas * 2);
+                    tracing::debug!(?err, "transaction already known");
+                } else {
+                    tracing::warn!(?err, "submission failed");
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+fn calculate_submission_gas_price(
+    gas_price_of_last_submission: Option<U256>,
+    web3_gas_estimation: GasPrice1559,
+    newest_nonce: U256,
+    nonce_of_last_submission: Option<U256>,
+) -> Result<GasPrice> {
+    let GasPrice1559 {
+        max_fee_per_gas, ..
+    }: gas_estimation::GasPrice1559 = web3_gas_estimation;
+    // The gas price of the refund tx is the current prevailing gas price
+    // of the web3 gas estimation plus a buffer.
+    // Since we are using Eip1559 gas specification,
+    // we will only pay the buffer if it is used
+    let mut new_max_fee_per_gas = max_fee_per_gas * GAS_PRICE_BUFFER_FACTOR;
+    // If tx from the previous submission was not mined,
+    // we incease the gas price
+    if let Some(nonce_of_last_submission) = nonce_of_last_submission {
+        if nonce_of_last_submission == newest_nonce {
+            // Increase the gas price to last submission's
+            // gas price plus an additional 12 %
+            if let Some(previous_gas_price) = gas_price_of_last_submission {
+                new_max_fee_per_gas = f64::max(
+                    new_max_fee_per_gas,
+                    previous_gas_price.to_f64_lossy() * GAS_PRICE_BUMP,
+                );
+            }
+        }
+    }
+    let max_fee_per_gas = U256::from_f64_lossy(new_max_fee_per_gas);
+    let gas_price = GasPrice::Eip1559 {
+        max_fee_per_gas,
+        max_priority_fee_per_gas: U256::min(max_fee_per_gas, U256::from(MAX_PRIORITY_FEE_TIP)),
+    };
+    Ok(gas_price)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_calculate_submission_gas_price() {
+        // First case: previous tx was successful
+        let max_fee_per_gas = 4_000_000_000f64;
+        let web3_gas_estimation = GasPrice1559 {
+            base_fee_per_gas: 2_000_000_000f64,
+            max_fee_per_gas,
+            max_priority_fee_per_gas: 2_000_000_000f64,
```

Comment:
Nit: can you use values that aren't the same (as well at not the same as `MAX_PRIORITY_FEE_TIP`) so we see if number were misattributed?